go/vt/vtgate/engine/create_database.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2020 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package engine
+
+import (
+	"strings"
+
+	"vitess.io/vitess/go/sqltypes"
+	querypb "vitess.io/vitess/go/vt/proto/query"
+	"vitess.io/vitess/go/vt/vtgate/vindexes"
+)
+
+var _ Primitive = (*DropCreateDatabase)(nil)
+
+// DropCreateDatabase is just a container around custom database provisioning plugins
+// The default behaviour is to just return an error
+type DropCreateDatabase struct {
+	name, verb string
+	plugin     func() error
+	noInputs
+	noTxNeeded
+}
+
+// CreateDropCreateDatabase creates the engine primitive
+func CreateDropCreateDatabase(dbName, verb string, plugin func() error) *DropCreateDatabase {
+	return &DropCreateDatabase{
+		name:   dbName,
+		verb:   verb,
+		plugin: plugin,
+	}
+}
+
+// RouteType implements the Primitive interface
+func (c *DropCreateDatabase) RouteType() string {
+	return c.verb + " database"
+}
+
+// GetKeyspaceName implements the Primitive interface
+func (c *DropCreateDatabase) GetKeyspaceName() string {
+	return c.name
+}
+
+// GetTableName implements the Primitive interface
+func (c *DropCreateDatabase) GetTableName() string {
+	return ""
+}
+
+// Execute implements the Primitive interface
+func (c *DropCreateDatabase) Execute(VCursor, map[string]*querypb.BindVariable, bool) (*sqltypes.Result, error) {
+	err := c.plugin()
```

Comment:
As a plugin author, how could I get access to the vtgate instance that's calling me? For example, to mimic a real `create database`, I'd want to block until vtgate can actually see the new keyspace I've provisioned, so I'd need to ask vtgate, "can you see it yet?" Similarly for `drop database` I'd want to block until the keyspace is sufficiently gone such that `create database` with the same name would succeed. In addition to talking to vtgate, I might also need to talk directly to topo for something like that.

Alternatively, should we implement these blocking checks here in upstream Vitess so that each plugin doesn't have to figure it out separately?